working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1748,65 +1746,196 @@ types of all of the case expressions.
+In a switch statement or expression, multiple cases may share the same body:
 
 ```dart
 switch (obj) {
   case [int a, int b]:
   case {"a": int a, "b": int b}:
     print(a + b); // OK.
+}
+```
 
-  case [int a]:
-  case (String a): // Error.
-    break;
+This would normally be a name collision, but we make an exception to allow it
+when it doesn't cause any problems.
+
+When compiling the body shared by a set of cases with pattern variable sets `vs`
+(where a default case has an empty pattern variable set), the enclosing scope is
+a synthesized scope `s` defined as:
+
+1.  Let `n` be the union of the sets of names of all of the pattern variable
+    sets in `vs`.
+
+2.  For each name in `n`:
+
+    1.  If `n` is defined in every pattern variable set in `vs` and has the same
+        type and finality, then introduce `n` into `s` with that type and
+        finality.
+
+        If any of the corresponding variables in `vs` are promoted, calculate
+        the promoted type of the variable in `s` based on all of the promoted
+        types of `n` in the cases in the same way that promotions are merged at
+        join points.
+
+    2.  Else `n` is inconsistently defined by the cases. Introduce `n` into `s`
+        with a sentinel type `Bad`.
+
+Compile the body using that scope. It is a compile-time error if any identifier
+in the body resolves to a variable in `s` whose type is `Bad`.
+
+*Note that it is not a compile-time error for there to be inconsistently
+defined variables between cases. It's only an error to use them in the body.
+This enables patterns to define inconsistent variables that are only used by
+their respective guards:*
+
+```dart
+switch (obj) {
+  case [var a, int n] when n > 1:
+  case [var a, double n] when n > 1.0:
+  case [var a, String s] when s.isNotEmpty:
+    print(a);
 }
 ```
 
-This would normally be a name collision, but we make an exception to allow this.
-However, it is a compile-time error if all switch case patterns that share a
-body do not all define the exact same variables with the exact same types.
+*This example has no errors because the only variable used in the body, `a`, is
+defined consistently by all cases.*
 
-*Aside from this special case, note that since all variables declared by a
-pattern and its subpatterns go into the same scope, it is an error if two
-subpatterns declare a variable with the same name, unless the name is `_`.*
+At runtime, the enclosing scope is *only* the case scope of the case that
+matched. *This determines which specific variables in which case are actually
+bound at runtime. We can't use a single synthesized scope that is shared by all
+cases because closures in guard clauses should see each case as having its own
+variables:*
+
+```dart
+bool capture(void Function() callback, bool result) {
+  callback();
+  return result;
+}
+
+switch ([1]) {
+  case [int a] when capture(() => a++, false):
+  case [int a] when capture(() => a += 10, true):
+    print(n);
+}
+```
+
```

Comment:
Mention that:
```dart
case [int a]:
default:
   break;
```
treats `default` as a case with zero declared variables.
(We can disallow combining `default` with anything else. Until now it was useless, but with `when` clauses with side effect, it might actually be useful now. Also, `default:` and `case _:` should probably just work the same.)

Same for labels:
```dart
label:
case [int a]: {
  ...
}
case [num a]: continue label;
```
which also allows a non-binding way into the case body.

Reply:
Done.